pep-0484.txt

```diff
@@ -448,6 +448,63 @@ is not generic but implicitly inherits from ``Iterable[Any]``::
 Generic metaclasses are not supported.
 
 
+Scoping rules for type variables
+--------------------------------
+
+Type variables follow normal name resolution rules.
+However, there are some special cases in the static typechecking context:
+
+* A type variable could be inferred to be equal to different types in a block,
+  unless this block is a generic class definition block. In such case,
+  all uses of a type variable are bound to the type that
+  parameterizes the generic class. Example::
+
+    from typing import TypeVar, Generic
+
+    T = TypeVar('T')
+
+    def fun_1(x: T) -> T: ... # T here
+    def fun_2(x: T) -> T: ... # and here could be different, see below
+
+    class MyClass(Generic[T]):
+        def meth_1(self, x: T) -> T: ... # but T here
+        def meth_2(self, x: T) -> T: ... # and here are always the same
+
+    fun_1(1)   # This is OK, T is inferred to be int
+    fun_2('a') # This is aslo OK, now T is str
+
+    a = MyClass() # type: MyClass[int]
+    a.meth_1(1)   # OK
+    a.meth_2('a') # This is an error!
+
+* Free type variables should not appear in the bodies of generic functions.
+  As well, types of class attributes should not contain free type variables::
+
+    T = TypeVar('T')
+    S = TypeVar('S')
+    class Foo(Generic[T]):
+        def method(self, x: T, y: S) -> S: # This is an error!
```

Comment:
Actually, I also tend to think that it is OK, but @rwbarton in https://github.com/python/mypy/issues/1903 says that it should be prohibited (as I understand, the reason is to avoid errors with forgotten type variables in generic class declaration). Maybe I misunderstood him.
I removed this part almost completely, if necessary we could add it later.

I only left a restriction that prohibits "free" type variables in the bodies (see new version).